MinecraftClient/McClient.cs

```diff
@@ -1084,6 +1085,102 @@ public bool MoveTo(Location location, bool allowUnsafe = false, bool allowDirect
+
+            List<Location> listOfBlocks = xRange.SelectMany(x => yRange.SelectMany(y => zRange.Select(z => new Location(x, y, z)))).ToList();
+
+            // Get all blocks that are air, have an air block above and a solid block below (space for a player to fit in)
+            // and sort them after their distance to goalToApproach. Take 5 by default to avoid too much calculation power.
+            List<Location> listOfCloseLocations = listOfBlocks.Where(block => GetWorld().GetBlock(new Location(block.X, block.Y + 1, block.Z)).Type == Material.Air &&      // AIR
+                                                                            GetWorld().GetBlock(block).Type == Material.Air &&                                              // AIR
+                                                                            GetWorld().GetBlock(new Location(block.X, block.Y - 1, block.Z)).Type.IsSolid())                // SOLID
+                                                                .OrderBy(closeLocation => goalToApproach.Distance(closeLocation)).ToList();
+
+            // If a minimal distance to the goal was given, remove all locations that are too close.
+            if (minDistance>0) 
+            {
+                listOfCloseLocations = listOfCloseLocations.Where(block => block.Distance(goalToApproach) >= minDistance).ToList();
+            }
+
+            // Check how much work to do. Avoid too much effort.
+            if (listOfCloseLocations.Count < 1)
+            {
+                return false;
+            }
+            else if (listOfCloseLocations.Count == 1)
+            {
+                return MoveTo(listOfCloseLocations[0], allowUnsafe);
+            }
+            else
+            {
+                // Create several instances of CalculatePath() to find ways to the closest locations
+                List<Task<Queue<Location>>> listOfPathfindInstances = new List<Task<Queue<Location>>>();
+                CancellationTokenSource cts = new CancellationTokenSource();
+                CancellationToken ct = cts.Token;
+
+                Task pathfindSpawner = Task.Run(() =>
+                {
+                    // Using Semaphore to limit the amount of parallel running tasks
+                    using (SemaphoreSlim concurrencySemaphore = new SemaphoreSlim(pathfindingThreads))
+                    {
+                        foreach (Location closeLoc in listOfCloseLocations)
+                        {
+                            // Wait if there are too many parallel tasks
+                            concurrencySemaphore.Wait();
+
+                            // Create a Task for one path
+                            listOfPathfindInstances.Add(Task<Queue<Location>>.Run(() => Movement.CalculatePath(world, this.location, closeLoc, allowUnsafe)));
+
+                            // If a valid path was found, stop spawning new tasks and exit
+                            if (ct.IsCancellationRequested) { break; }

```

Comment:
If exiting early, maybe the method will end up selecting the first valid path, which happens to be the farthest one in the searched radius? To avoid that while still retaining performance (i.e. ability to stop early), how about sorting candidates first by `distanceSquared()` with the destination (which should be relatively fast to compute), then start calculating paths? So stopping early should still yield a good (if not best) result.